File: src/ts/ssh-tcp/messages/portForwardRequestMessage.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { SessionRequestMessage, SshDataReader, SshDataWriter } from '@microsoft/dev-tunnels-ssh';
import { PortForwardingService } from '../services/portForwardingService';

export class PortForwardRequestMessage extends SessionRequestMessage {
	public addressToBind: string = '';
	public port: number = 0;

	public constructor() {
		super();
		this.requestType = PortForwardingService.portForwardRequestType;
		this.wantReply = true;
	}

	protected onRead(reader: SshDataReader): void {
		super.onRead(reader);

		this.addressToBind = reader.readString('ascii');
		this.port = reader.readUInt32();
	}

	protected onWrite(writer: SshDataWriter): void {
		super.onWrite(writer);

		writer.writeString(this.validateField(this.addressToBind, 'address'), 'ascii');
		writer.writeUInt32(this.validateField(this.port, 'port'));
	}

	public toString() {
		return `${super.toString()} (addressToBind=${this.addressToBind} port=${this.port})`;
	}
}

File: src/cs/Ssh.Tcp/Messages/PortForwardRequestMessage.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Text;
using Microsoft.DevTunnels.Ssh.IO;
using Microsoft.DevTunnels.Ssh.Tcp;

namespace Microsoft.DevTunnels.Ssh.Messages;

public class PortForwardRequestMessage : SessionRequestMessage
{
	public string? AddressToBind { get; set; }

	public uint Port { get; set; }

	public PortForwardRequestMessage()
	{
		RequestType = PortForwardingService.PortForwardRequestType;
		WantReply = true;
	}

	protected override void OnRead(ref SshDataReader reader)
	{
		base.OnRead(ref reader);
		AddressToBind = reader.ReadString(Encoding.ASCII);
		Port = reader.ReadUInt32();
	}

	protected override void OnWrite(ref SshDataWriter writer)
	{
		base.OnWrite(ref writer);
		writer.Write(AddressToBind ?? string.Empty, Encoding.ASCII);
		writer.Write(Port);
	}

	public override string ToString()
	{
		return base.ToString() + $"(AddressToBind: {AddressToBind}, Port: {Port})";
	}
}
